### PR TITLE
Implement #3129 (Optimized history traversals in the REST API)

### PR DESCRIPTION
**Pull request by zah**
Up to 10x speed-up for historic traversals through the API:

```
# Before
➜  nimbus-eth2 git:(feature/rest_api_optimized_history_traversal) build/rest_api_benchmark --start-slot=9000 --count=250
INF 2021-12-22 09:24:55.962+02:00 Benchmarking ...                           file=rest_api_benchmark.nim:42 apiName=getStateRoot
INF 2021-12-22 09:26:08.637+02:00 getStateRoot                               file=rest_api_benchmark.nim:43 t=72.67443995799999s
INF 2021-12-22 09:26:08.637+02:00 Benchmarking ...                           file=rest_api_benchmark.nim:42 apiName=getStateFork
INF 2021-12-22 09:27:20.277+02:00 getStateFork                               file=rest_api_benchmark.nim:43 t=71.639665333s
INF 2021-12-22 09:27:20.277+02:00 Benchmarking ...                           file=rest_api_benchmark.nim:42 apiName=getStateFinalityCheckpoints
INF 2021-12-22 09:28:33.398+02:00 getStateFinalityCheckpoints                file=rest_api_benchmark.nim:43 t=73.12129654100001s
INF 2021-12-22 09:28:33.398+02:00 Benchmarking ...                           file=rest_api_benchmark.nim:42 apiName=getStateValidatorBalances
INF 2021-12-22 09:29:55.869+02:00 getStateValidatorBalances                  file=rest_api_benchmark.nim:43 t=82.470523375s

# After
➜  nimbus-eth2 git:(feature/rest_api_optimized_history_traversal) build/rest_api_benchmark --start-slot=9000 --count=250
INF 2021-12-22 09:24:55.962+02:00 Benchmarking ...                           file=rest_api_benchmark.nim:42 apiName=getStateRoot
INF 2021-12-22 09:26:08.637+02:00 getStateRoot                               file=rest_api_benchmark.nim:43 t=6.67443995799999s
INF 2021-12-22 09:21:32.563+02:00 Benchmarking ...                           file=rest_api_benchmark.nim:42 apiName=getStateFork
INF 2021-12-22 09:21:39.626+02:00 getStateFork                               file=rest_api_benchmark.nim:43 t=7.062567791s
INF 2021-12-22 09:21:39.626+02:00 Benchmarking ...                           file=rest_api_benchmark.nim:42 apiName=getStateFinalityCheckpoints
INF 2021-12-22 09:21:46.180+02:00 getStateFinalityCheckpoints                file=rest_api_benchmark.nim:43 t=6.55376425s
INF 2021-12-22 09:21:46.180+02:00 Benchmarking ...                           file=rest_api_benchmark.nim:42 apiName=getStateValidatorBalances
INF 2021-12-22 09:22:01.645+02:00 getStateValidatorBalances                  file=rest_api_benchmark.nim:43 t=15.465585083s
```

### CODE REVIEW

File: docs/the_nimbus_book/src/options.md [hunk 1]
```diff
@@ -21,89 +21,75 @@ nimbus_beacon_node [OPTIONS]... command
 
 The following options are available:
 
-     --log-level               Sets the log level for process and topics (e.g. "DEBUG;
-                               TRACE:discv5,libp2p; REQUIRED:none; DISABLED:none") [=INFO].
-     --log-format              Specifies what kind of logs should be written to stdout (beta)
-                               (auto, colors, nocolors, json) [=auto].
-     --log-file                Specifies a path for the written Json log file.
+     --log-level               Sets the log level for process and topics (e.g. "DEBUG; TRACE:discv5,libp2p;
```
Comment (arnetheduck) [inline docs/the_nimbus_book/src/options.md:24]: the help text here depends on the width of the terminal you generate it on .. afair it's something in the 90-100 character range, to avoid these unnecessary changes

File: beacon_chain/spec/eth2_apis/rest_types.nim [hunk 2]
```diff
@@ -219,6 +219,7 @@ type
 
   RestMetadata* = object
     seq_number*: string
+    syncnets*: string
```
Comment (arnetheduck) [inline beacon_chain/spec/eth2_apis/rest_types.nim:222]: afair not all servers serve this field - it would need to be read with lax rules

File: docs/the_nimbus_book/src/options.md (hunk 1 above)
Comment (zah) [reply docs/the_nimbus_book/src/options.md:24]: I'm aware of this. There is a maximum width in Confutils and I think we should stick to using it. Otherwise, how is everyone supposed to find out the width used by the previous developer when generating the help text?

Comment (arnetheduck) [reply docs/the_nimbus_book/src/options.md:24]: well, write it down here, in a markdown comment? the point is to avoid a battle of the widths-scenario where every time each developer uses their favorite setting.. I don't _really_ care what the width is, except that it should fit on a reasonable browser.

wdyt of a `make update-help` command that is run as part of the build every time `conf.nim` changes?

File: beacon_chain/conf.nim [hunk 3]
```diff
@@ -327,6 +327,16 @@ type
         defaultValueDesc: "127.0.0.1"
         name: "rest-address" }: ValidIpAddress
 
+      restTraversals* {.
```
Comment (arnetheduck) [inline beacon_chain/conf.nim]:
```suggestion
      restCacheSize* {.
```

I think calling this "traversals" unnecessarily limits the feature and abstracts it from what actually is going on in a slightly misleading way - ie traversals is _one_ operation that gets faster, but so does re-querying the same info over and over - later, once can imagine expanding the use cases in other ways as the cache and db gets smarter

File: beacon_chain/rpc/rest_traversals_cache.nim [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+logScope:
+  topics = "rest_traversals_cache"
+
+proc init*(T: type RestTraversalsCache,
+           cacheSize: Natural,
+           cacheTtl: Duration): T =
+  doAssert cacheSize > 0
+
+  RestTraversalsCache(
```
Comment (arnetheduck) [inline beacon_chain/rpc/rest_traversals_cache.nim]: ditto - `StateTtlCache` maybe?

File: beacon_chain/rpc/rest_utils.nim [hunk 5]
```diff
@@ -116,19 +116,50 @@ proc getBlockDataFromBlockIdent*(node: BeaconNode,
...
+        node.restTraversalsCache.getClosestState(blockSlot.slot)
+      else:
+        nil
+
+      let stateToAdvance = if cachedTraversalState != nil:
+        cachedTraversalState
+      else:
+        assignClone(node.dag.headState)
+
+      node.dag.updateStateData(stateToAdvance[], blockSlot, false, cache)
```
Comment (arnetheduck) [inline beacon_chain/rpc/rest_utils.nim:182]: as of https://github.com/status-im/nimbus-eth2/pull/3217, this can fail - there are also some conflicting fixes to the rest api in that branch, as all state requests, as of the checkpoint feature, can fail

File: beacon_chain/rpc/rest_traversals_cache.nim [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  var
+    bestSlotDifference = Slot.high
+    index = -1
+
+  for i in 0 ..< cache.entries.len:
+    if cache.entries[i] == nil:
+      continue
+
+    let stateSlot = getStateField(cache.entries[i].state.data, slot)
+    if stateSlot > slot:
```
Comment (arnetheduck) [inline beacon_chain/rpc/rest_traversals_cache.nim]: see https://github.com/status-im/nimbus-eth2/blob/1a6a56bdb1aab0b1778e99cd1c5834180c4a52ca/beacon_chain/consensus_object_pools/blockchain_dag.nim#L867 - this condition is both not enough and too strong: an exact slot match is fine as long as the block matches (the block has not been applied at the given slot), but you also need to check that the state belongs to the same history in the dag - this is important in two cases: 
* a state id was used to point to an explicit "alternative" history
* a reorg happened between two requests and the "default" canonical history most often used in rest requests changed

the above is covered mostly by the `canAdvance` construct in `updateStateData`, which in the loop ensures that the correct history is being considered (in `ChainDAG`) - the loop in which it is used performs the same logic as is desired here.

**Comment by arnetheduck** [comment]
I've keep wondering why `Implement #3129` as the main headline for a git commit makes sense - in nim-lang, this is also the culture, what's the .. idea? 

The log message is written _for_ the consumer of it, and when reading the history (`git log --oneline`), 3129 isn't meaningful / helpful - it seems to me that `Bla bla (implements #XXXX)` is more useful to readers?

File: beacon_chain/conf.nim [hunk 7]
```diff
@@ -327,6 +327,16 @@ type
...
+        defaultValue: 10
```
Comment (arnetheduck) [inline beacon_chain/conf.nim]: this is actually quite a heavy default - ie it's roughly 100mb per state afair - we might want to start with something like 1 or 2 that roughly corresponds to the current ceiling consumption of the REST component

File: beacon_chain/conf.nim (hunk 3 above)
Comment (zah) [reply beacon_chain/conf.nim]: Re-querying the same info doesn't get faster because we store the advanced state in the cache (traversals won't be really speed-up otherwise).

File: beacon_chain/rpc/rest_utils.nim (hunk 5 above)
Comment (zah) [reply beacon_chain/rpc/rest_utils.nim:182]: This PR is simpler though, so it makes sense to merge it first perhaps. #3217 haven't been reviewed yet.

File: beacon_chain/conf.nim (hunk 7 above)
Comment (zah) [reply beacon_chain/conf.nim]: If you don't perform alternative traversals, you don't pay for anything. Also, the cache expires very quickly. The limit is put there only to ensure that this cannot be used as a DoS vector against the server (even though we know that the REST API is an effective DoS vector anyway). Putting the cap at 1GB doesn't look that high to me. Using lower values is certainly reasonable, but 1 or 2 is rather extreme IMO. Perhaps 3 to 5 is okay.

File: docs/the_nimbus_book/src/options.md (hunk 1 above)
Comment (zah) [reply docs/the_nimbus_book/src/options.md:24]: This is a neat idea, but someone must spend the time implementing it.

File: beacon_chain/rpc/rest_traversals_cache.nim (hunk 6 above)
Comment (zah) [reply beacon_chain/rpc/rest_traversals_cache.nim]: `canAdvance` doesn't check the right conditions. It checks that the state can be fast-forwarded by applying empty slots (it doesn't have intermediate blocks in between). 

The modification in chaindag made in this PR enables advancing even in the presence of intermediate blocks (they will be discovered in the loop gathering blocks that used to terminate at a database snapshot). 

You are right that checking that we are on the same branch in the history is necessary as well, but it's not necessary for correctness, only to ensure that the optimization actually brings benefits. Considering that this cache is intended for historic queries where alternative histories are not present, being overly strict in the conditions may be an overkill.

File: beacon_chain/conf.nim (hunk 3 above)
Comment (zah) [reply beacon_chain/conf.nim]: Actually, re-querying can be faster indeed. I'll make the suggested change.

File: docs/the_nimbus_book/src/options.md (hunk 1 above)
Comment (arnetheduck) [reply docs/the_nimbus_book/src/options.md:24]: well, it's a matter of running the compiled binary, there's nothing to implement really - this problem of resizing the help text to different widths comes up pretty much every time I review help text updates - it's a bit of a churn - even without any automation, just using a simple redirect of `--help` should be enough - that'll automatically use whatever width confutils uses for non-terminals.

File: beacon_chain/rpc/rest_traversals_cache.nim (hunk 6 above)
Comment (arnetheduck) [reply beacon_chain/rpc/rest_traversals_cache.nim]: > It checks that the state can be fast-forwarded by applying empty slots (it doesn't have intermediate blocks in between).

hm, taken together with the loop it runs in, it _should_ check for the right condition - the conditions are exactly the same for this cache and for the `updateStateData` - we should be checking that histories match and that slots are reasonable - in the case of `updateStateData` it's necessary for correctness - in the case of this cache, it's necessary to reach optimal efficiency - the logic is exactly the same.

Comment (arnetheduck) [reply beacon_chain/rpc/rest_traversals_cache.nim]: > Considering that this cache is intended for historic queries 

in the case of non-finality, historic queries quickly become "current queries" - ie there's no guarantee that all callers of this feature use it for historic queries - it should nonetheless reuse a reasonable cache entry if it has it - for example, it's likely calls to the rest API query the state at a given offset from head - say.. `head-6` slots - this gives callers the ability to reuse the state without loading from DB when following head at a "safe" distance that's likely to become final, similar to how exchanges today require N verifications.

File: beacon_chain/conf.nim (hunk 3 above)
Comment (arnetheduck) [reply beacon_chain/conf.nim]: Not only that: slot-by-slot queries that follow head at a fixed distance also become faster etc - there is nothing inherently limiting this feature to historical queries even if that is how it was conceived in this PR - a general cache like this has been part of the REST API plan for some time.

File: beacon_chain/conf.nim [hunk 8]
```diff
@@ -327,6 +327,16 @@ type
...
+        desc: "The maximum number of expected concurrent historic traversals (API requests obtaining consecutive slot or epoch information)"
+        name: "rest-traversals" }: Natural
+
+      restTraversalsTtl* {.
+        defaultValue: 60
```
Comment (arnetheduck) [inline beacon_chain/conf.nim:339]: this should probably be either the duration of "a few slots+some" or maybe an "epoch + some" so that queries at "regular" intervals tied to the constants in the spec "make sense"

File: beacon_chain/consensus_object_pools/blockchain_dag.nim [hunk 9]
```diff
@@ -936,8 +937,10 @@ proc updateStateData*(
 
     # Look for a state in the database and load it - as long as it cannot be
     # found, keep track of the blocks that are needed to reach it from the
-    # state that eventually will be found
-    while not dag.getState(state, cur):
+    # state that eventually will be found. Besides finding the state in the
+    # database we may also reach the input state provided to the function.
+    # It can also act as a viable starting point for the block replay later.
+    while cur.blck.root != state.blck.root and not dag.getState(state, cur):
```
Comment (arnetheduck) [inline beacon_chain/consensus_object_pools/blockchain_dag.nim]: I don't quite follow this change - we've already verified above that the starting point is _not_ the right block - `getState` will only return true when the `cur.blck.root` == `state.blck.root` so when exactly is the extra check against `cur` needed?

Comment (arnetheduck) [reply beacon_chain/consensus_object_pools/blockchain_dag.nim]: also, this will fail when state has been advanced _past_ the point of interest but matches block - ie if `bs.blck` is from slot 42 and we want to load `(blck, 100)`, we might have a `state` that's been advanced to `(blck, 105)` that falsely would match the new condition

Comment (arnetheduck) [reply beacon_chain/consensus_object_pools/blockchain_dag.nim]: ok, I think I understand when this triggers - it's the case where the given state is older than whatever we look for above in the `canAdvance` loop but newer than whatever is in the database - just like everything else, this needs to use a `canAdvance` check to be correct

File: beacon_chain/rpc/rest_traversals_cache.nim (hunk 6 above)
Comment (arnetheduck) [reply beacon_chain/rpc/rest_traversals_cache.nim]: > being overly strict

also, checking the history as well is not "overly strict" - it will never result in any false negatives - it will simply increase the efficiency of the cache in certain cases (because it'll select a better entry right when it matters the most: when the network is already under load)

File: beacon_chain/conf.nim (hunk 7 above)
Comment (arnetheduck) [reply beacon_chain/conf.nim]: well, you do pay: before this branch, querying non-historical, non-head states would have a cap of 1 BeaconState "extra memory used" no matter how many queries you made, in which order, and for what data - it could be very slow, but it would never consume more memory than that - now that cap goes up to whatever this setting is, for _general_ use of REST

File: beacon_chain/rpc/rest_traversals_cache.nim [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  chronos,
+  chronicles,
+  ../consensus_object_pools/block_pools_types
+
+type
+  CacheEntry = ref object
+    state: ref StateData
+    lastUsed: Moment
+
+  RestTraversalsCache* = ref object
```
Comment (arnetheduck) [inline beacon_chain/rpc/rest_traversals_cache.nim]: this looks like it doesn't really have to be `ref` - it can simply have 0 entries if the cache is disabled which simplifies the code quite a bit - it can simply be made to work correctly with 0 entries which will get rid of potential nils, doasserts and so on

File: docs/the_nimbus_book/src/options.md (hunk 1 above)
Comment (zah) [reply docs/the_nimbus_book/src/options.md:24]: There will be some new Makefiles rules, there will be a solution for the problem of splicing the help output in the book, etc (is there something like `include` statement in mdBook?). I don't see a reason for postponing merging this PR until the suggestion is implemented.

File: beacon_chain/rpc/rest_utils.nim (hunk 5 above)
Comment (arnetheduck) [reply beacon_chain/rpc/rest_utils.nim:182]: ok - in that case, let's remove the change in `blockchain_dag` from this PR which is incorrect afaict (and conflicting) - we can add back a correct version when both PR:s have made it through (or add it to the other PR)

File: docs/the_nimbus_book/src/options.md (hunk 1 above)
Comment (arnetheduck) [reply docs/the_nimbus_book/src/options.md:24]: it's not a blocker indeed, but an obvious low-hanging fruit

File: beacon_chain/conf.nim (hunk 7 above)
Comment (zah) [reply beacon_chain/conf.nim]: How about 3 then?

Comment (arnetheduck) [reply beacon_chain/conf.nim]: sure why not, it's not a bad number at all: https://en.wikipedia.org/wiki/3

File: beacon_chain/conf.nim (hunk 8 above)
Comment (zah) [reply beacon_chain/conf.nim:339]: An epoch is almost 7 minutes though, which is quite a significant deviation from this suggested default. If your requests are so rare, perhaps you won't benefit significantly from the cache. And indeed, using a larger TTL will put more weight on the suggestion for reducing the default cache size.

Comment (arnetheduck) [reply beacon_chain/conf.nim:339]: yeah indeed - I'm thinking about the case where there's roughly one query per epoch (which feels kind of reasonable as far as use cases go, given this is how often the effective balance, proposers and so on change)

File: beacon_chain/rpc/rest_utils.nim (hunk 5 above)
Comment (zah) [reply beacon_chain/rpc/rest_utils.nim:182]: I'm not sure I understand. We cannot remove the change in `blockchain_dag` as it is crucial for the optimization brought by this PR.

Comment (arnetheduck) [reply beacon_chain/rpc/rest_utils.nim:182]: well, it's wrong (it'll cause crashes) and conflicting (3217 adds error handling to this loop) so my idea was to remove it from this PR, then merge this PR, then introduce 3217 which should go in soon (it fixes a critical bug in unstable among other things), then we can make a separate change or change 3217 to include the (corrected) change. it's a matter of merge logistics, mostly.

Comment (zah) [reply beacon_chain/rpc/rest_utils.nim:182]: Well, it's also trivial to fix, but perhaps I'm misunderstanding. Are you referring to the new terminating condition in the block gathering loop in `updateStateData`? It's also wrong only in the sense that it can be misused by other API users in the future. Can you trigger a crash with the REST API as implemented in this PR?

Comment (arnetheduck) [reply beacon_chain/rpc/rest_utils.nim:182]: > new terminating condition

yes

Comment (arnetheduck) [reply beacon_chain/rpc/rest_utils.nim:182]: but whatever..

**Comment by arnetheduck** [comment]
fwiw, most comments on the PR are of the "value-on-the-table" type, except the loop condition thing - with that fixed, we can merge it (but it would be a shame, the comments will end up forgotten)

File: beacon_chain/rpc/rest_utils.nim (hunk 5 above)
Comment (arnetheduck) [reply beacon_chain/rpc/rest_utils.nim:182]: I put the correct change in 3217 as well so that it doesn't get lost in any merge order

File: beacon_chain/rpc/rest_traversals_cache.nim (hunk 10 above)
Comment (zah) [reply beacon_chain/rpc/rest_traversals_cache.nim]: Needs to be a ref to facilitate capturing it in a closure later in this module. Comment added.

Comment (arnetheduck) [reply beacon_chain/rpc/rest_traversals_cache.nim]: for the other managers and pools, we've stopped using `ref object` and instead use `ref Xxx` where a reference is needed for async - doing so allows keeping the code tight with respect to `var` vs `non-var` access in functions - I'm planning on giving `ChainDAG` this treatment as well: this would prevent stuff like REST from mutating the caches when it shouldn't, for example.

I wish we had `ref-to-immutable-instance` - instead we got the really weird strict func which is a really odd and inflexible way of looking at the problem :/

File: beacon_chain/rpc/rest_utils.nim [hunk 11]
```diff
@@ -116,19 +116,50 @@ proc getBlockDataFromBlockIdent*(node: BeaconNode,
...
-    let rpcState = assignClone(node.dag.headState)
-    node.dag.withState(rpcState[], blockSlot):
-      body
+    template isState(state: StateData): bool =
+      state.blck.atSlot(getStateField(state.data, slot)) == blockSlot
+
+    var cache {.inject, used.}: StateCache
+
+    # TODO view-types
+    # Avoid the code bloat produced by the double `body` reference through a lent var
```
Comment (arnetheduck) [inline beacon_chain/rpc/rest_utils.nim:167]: actually, a reasonable thing to do here might be to introduce a `proc runBody(state: StateData...)` - that would guarantee that the body doesn't mutate the state

Comment (zah) [reply beacon_chain/rpc/rest_utils.nim:167]: How would that work? `body` is a block in the caller scope. If you put it in a closure, you'll start getting all kinds of additional restrictions regarding what variables can be referenced within the body. The codegen will be different because some local variables will be moved to the heap, parameters may end up being copied, etc.

Comment (arnetheduck) [reply beacon_chain/rpc/rest_utils.nim:167]: yes, there are downsides :)

but yeah, view types (well.. immutable references with bounded lifetimes) would be the proper solution here - we should in fact also use clearancestate/epochrefstate, _and_ we should ensure that inside `body`, there are no await statements and other escapes.

File: beacon_chain/rpc/rest_utils.nim [hunk 12]
```diff
@@ -116,19 +116,50 @@ proc getBlockDataFromBlockIdent*(node: BeaconNode,
...
+
+      node.dag.updateStateData(stateToAdvance[], blockSlot, false, cache)
+
+      let providedState = if cachedState == nil and node.stateTtlCache != nil:
+        # This was not a cached state, we can cache it now
+        node.stateTtlCache.add(stateToAdvance)
+        stateToAdvance
+      else:
+        # If we have a cache hit, there is a concern that the REST request
+        # handler may continue executing asynchronously while we hit the same
```
Comment (arnetheduck) [inline beacon_chain/rpc/rest_utils.nim]: this must never happen - if we use `headState` with such a `body`, it will break because the head state can change while the body is running

**Review by arnetheduck — APPROVED** [review]

File: beacon_chain/rpc/rest_utils.nim (hunk 11 above)
Comment (zah) [reply beacon_chain/rpc/rest_utils.nim:167]: It can be done with my `fauxClosure` magic, but I have some doubts that you would like this :P

A `fauxClosure` is something like a closure that must be executed immediately. What the magic does is to turn every captured variable into a regular parameter of a non-closure proc with the supplied body and then this proc is immediately called. I would say that there is a rather low risk of this producing invalid code - if there are bugs they are likely to result in compilation problems (either in Nim or in C). The mechanism has been used extensively by the status-desktop team.

Besides reducing the code bloat, the generated procedure will be effectively non-async, eliminating the yielding risk.

File: beacon_chain/rpc/rest_utils.nim (hunk 12 above)
Comment (zah) [reply beacon_chain/rpc/rest_utils.nim]: Partial "no awaits" enforcement was implemented here:
https://github.com/status-im/nimbus-eth2/pull/3219/commits/524912fba58dff5020594666637dd961ee1b3f0f

File: beacon_chain/consensus_object_pools/blockchain_dag.nim [hunk 13]
```diff
@@ -609,6 +609,7 @@ proc getEpochRef*(dag: ChainDAGRef, blck: BlockRef, epoch: Epoch): EpochRef =
   let
     ancestor = epochAncestor(blck, epoch)
 
+  dag.epochRefState.blck = BlockRef()
```
Comment (arnetheduck) [inline beacon_chain/consensus_object_pools/blockchain_dag.nim:612]: what's this?

Comment (zah) [reply beacon_chain/consensus_object_pools/blockchain_dag.nim:612]: Since the state here is always an "output parameter" in intent, this nullifies the possibilities that it will be used as an input parameter due to the new change. At least some of the logic in the tests have to be changed otherwise.

Comment (arnetheduck) [reply beacon_chain/consensus_object_pools/blockchain_dag.nim:612]: but .. it should be used as an input parameter? basically, if replaying from epochrefstate is the best thing we can do, why wouldn't we?

Comment (arnetheduck) [reply beacon_chain/consensus_object_pools/blockchain_dag.nim:612]: not needed after all, will fix in error handling branch
